wgpu-hal/src/lib.rs

```diff
@@ -1242,8 +1242,38 @@ pub trait CommandEncoder: WasmNotSendSync + fmt::Debug {
 
     // pass common
 
-    /// Sets the bind group at `index` to `group`, assuming the layout
-    /// of all the preceding groups to be taken from `layout`.
+    /// Sets the bind group at `index` to `group`.
+    ///
+    /// If this is not the first call to `set_bind_group` within the current
+    /// render or compute pass:
+    ///
+    /// - If `layout` contains `n` bind group layouts, then any previously set
+    ///   bind groups at indices `n` or higher are cleared.
+    ///
+    /// - If the first `m` bind group layouts of `layout` are equal to those of
+    ///   the previously passed layout, but no more, then any previously set
+    ///   bind groups at indices `m` or higher are cleared.
```

Comment:
Do all native APIs we target clear all `n`/`m` or higher bind groups?

I thought only Vulkan required users to rebind the `n`/`m` or higher bind groups.

Reply:
`wgpu_hal` is supposed to be a portability layer, so can only promise what the weakest implementation can deliver. We could place the burden on Vulkan to track more state, but this seems like the kind of work that is best addressed in `wgpu_core`.

But the punch line is, unless I'm misunderstanding things, Metal actually needs this requirement too. Each MSL entry point refers to all its resources via a flat numbering scheme (`[[texture(0)]]`, `[[texture(1)]]`, `[[texture(2)]]`), with no analogue of WGSL's `@group` / `@binding` two-level scheme. At the Metal API level, we just call `[encoder setTexture:t atIndex:2]`. We assign MSL indices consecutively, so the indices of resources supplied by higher-numbered bind groups are influenced by the number of those resources present in lower-numbered bind groups. This means that, after we switch pipelines, the MSL numbering of resources supplied by higher-numbered bind groups could be off. So MSL effectively has the same requirement, arising from our desire to densely number resources.

Reply:
Note that if we were to switch `wgpu_hal::metal` to use argument buffers, Metal would no longer have this restriction, because this would actually introduce this two-level numbering: the argument buffers themselves would have `[[buffer(g)]]` indices corresponding to `@group(g)` indices, and the `[[id(b)]]` attributes within the argument buffer would correspond to `@binding(b)` indices. Then higher-level bind groups could be stable across pipeline layout changes.

Reply:
However, `wgpu_hal` still probably shouldn't promise stability, since there isn't a natural way to implement it on Vulkan. So the above doesn't serve as a reason to prioritize using argument buffers on Metal.

Reply:
> `wgpu_hal` is supposed to be a portability layer, so can only promise what the weakest implementation can deliver. We could place the burden on Vulkan to track more state, but this seems like the kind of work that is best addressed in `wgpu_core`.

I agree; it wasn't my intention to put the burden on Vulkan to track more state.

My comment was targeted at the word "cleared". If not all backends "clear", then `wgpu-hal` shouldn't promise that.

I think we should instead require users of the API to rebind the bind groups but not promise that they will be cleared.

Reply:
I see what you're saying. I think I agree with you about what `wgpu_hal` should promise - but I still think there's a piece we're missing. However, I want a better place for the discussion than this issue; I'll bring it up in chat.